workflow/controller/exit_handler.go

```diff
@@ -15,7 +15,7 @@ import (
 
 func (woc *wfOperationCtx) runOnExitNode(ctx context.Context, exitHook *wfv1.LifecycleHook, parentNode *wfv1.NodeStatus, boundaryID string, tmplCtx *templateresolution.Context, prefix string, scope *wfScope) (bool, *wfv1.NodeStatus, error) {
 	outputs := parentNode.Outputs
-	if parentNode.Type == wfv1.NodeTypeRetry {
+	if parentNode.Type == wfv1.NodeTypeRetry && !(parentNode.MemoizationStatus != nil && parentNode.MemoizationStatus.Hit) {
```

Comment:
This is a complex thing to get your head around (not that it wasn't before your change). I agree it does the right thing, but I'd like to wrap up this if and the subsequent `getChildNodeIndex()` call into a single helper function.
It feels like it may result in breakage in the future unless we do this.

```
// Check if we have a retry node which wasn't memoized and return that if we do
func (woc *wfOperationCtx) possiblyGetRetryChildNode(node *wfv1.NodeStatus) *wfv1.NodeStatus {
	if node.Type == wfv1.NodeTypeRetry && !(node.MemoizationStatus != nil && node.MemoizationStatus.Hit) {
		return getChildNodeIndex(node, woc.wf.Status.Nodes, -1)
	}
	return nil
}
```

and then this becomes
```
if lastChildNode := possiblyGetRetryChildNode(node); lastChildNode != nil {
  outputs = lastChildNode.Outputs
}
```

Reply:
Thanks! I agree with you！ I have changed all.